### ISSUE TITLE
SDKs should expose history length and size via Workflow info

### ISSUE THREAD

**Issue by bergundy** [temporalio/features#16]
History length:

- Golang - https://github.com/temporalio/sdk-go/pull/851
- TS - https://github.com/temporalio/sdk-typescript/issues/357
- Java - https://github.com/temporalio/sdk-java/issues/1300
- Python - https://github.com/temporalio/sdk-python/pull/73
- PHP - https://github.com/temporalio/sdk-php/issues/222
- .NET - https://github.com/temporalio/sdk-dotnet/issues/122

History size and "should continue as new":

- Server - https://github.com/temporalio/temporal/issues/2726 (https://github.com/temporalio/temporal/pull/3055)
- Golang - https://github.com/temporalio/sdk-go/issues/628
- Core - https://github.com/temporalio/sdk-core/issues/564
- TS - https://github.com/temporalio/sdk-typescript/issues/695
- Java - https://github.com/temporalio/sdk-java/pull/1853
- Python - https://github.com/temporalio/sdk-python/pull/361
- PHP - TODO
- .NET - https://github.com/temporalio/sdk-dotnet/issues/122

**Comment by lorensr** [comment]
Related:

> We should give `Workflow.isContinueAsNewNeeded` flag to the SDKs. And it should be coming from the service.

https://temporaltechnologies.slack.com/archives/C01FT8U10GK/p1643400086752909?thread_ts=1643398281.925129&cid=C01FT8U10GK

**Comment by cretz** [comment]
Should also expose the history size and not just length. See https://github.com/temporalio/api/pull/178 (but unsure if server-populated yet).

**Comment by lorensr** [comment]
I think we should expose `shouldContinueAsNew` as well in all SDKs.

1. If server is [able to take into account other things](https://github.com/temporalio/temporal/issues/4117), then it's better than history length/size
2. This:

```ts
if (workflowInfo().shouldContinueAsNew) {
  cAN()
}
```
is shorter & easier to read/understand compared to:
```ts
if (workflowInfo().historyLength < 10_000 && workflowInfo().historySize < 10_000_000) {
  cAN()
}
```
3. We'll no longer need to teach people the right numbers to use

**Comment by cretz** [comment]
> I think we should expose `shouldContinueAsNew` as well in all SDKs.

Concur, when available in server (no benefit in supporting before server)

**Comment by mjameswh** [comment]
It will be important to mention in documentation the fact that relying on these properties is useless and dangerous if using an older server.

**Comment by cretz** [comment]
We could make it a capability and error (i.e. wft failure) if invoked for older server. But I think this is a good enough reason to not even expose this until it is available _on cloud_.

**Comment by cretz** [comment]
A caveat that SDKs need to note here: these three values may be delayed if viewed via query. Since the value is on wft start, a query-only poll response won't have updated history. So if you, say, sent a signal to start far-out 50 timers, and then, even after already in history stored, sent just a query, there is no new wft and therefore no value updates even though current event count is actually quite different on the server.

Basically every SDK needs to document that these values may be delayed if accessed in a query.
